src/python/WMCore/MicroService/MSTransferor/MSTransferor.py

```diff
@@ -676,3 +690,132 @@ def _diskPNNs(self, pnnList):
             else:
                 diskPNNs.add(pnn)
         return diskPNNs
+
+    def updateSites(self, doc):
+        """
+        Update sites API provides asynchronous update of Site info.
+
+        :param doc: JSON payload with the following data structures:
+                    [{'workflow': <wflow name>, 'SiteWhiteList' ['T1', ...], 'SiteBlackList': ['T2',...]}, {...}]
+        :return: acknowledge dict to upstream caller (ReqMgr2)
+        """
+        # preserve provided payload to local file system
+        errors = []
+        for rec in doc:
+            wflow = rec['workflow']
+            status = self.saveData(wflow, rec)
+            if status != 'ok':
+                errors.append({'workflow': wflow, 'error': status})
+        # send acknowledged message back to upstream caller
+        resp = {'status': 'ok'}
+        if len(errors) != 0:
+            resp = {'status': 'fail', 'errors': errors}
+        return resp
+
+    def saveData(self, wflow, rec):
+        """
+        Save workflow data to persistent storage
+        :param wflow: name of workflow
+        :param rec: record to save
+        :return: status of this operation
+        """
+        try:
+            fname = '{}/{}'.format(self.storage, wflow)
+            with open(fname, 'w', encoding="utf-8") as ostream:
+                ostream.write(rec)
+            return 'ok'
+        except Exception as exp:
+            msg = "Unable to save workflow '%s' to %s.Error: %s", wflow, self.storage, str(exp)
+            self.logger.exception(msg)
+            return str(exp)
+
+    def readData(self, wflow):
+        """
+        Read workflow data from persistent storage
+        :return: data or None
+        """
+        fname = '{}/{}'.format(self.storage, wflow)
+        data = None
+        with open(fname, 'r', encoding="utf-8") as istream:
+            data = json.load(istream.read())
+        return data
+
+    def _updateSites(self, wflow):
```

Comment:
I would remove all this code and rely on what has already been implemented in MSTransferor, hence, just let the workflow go through the standard algorithm.

When removing this module, please do not squash commits though. Just in case I am missing any detail that would make that not possible.

Reply:
I'm not sure why I need to remove it since it is a business logic of requested feature. How standard algorithm will execute a logic which is not there? So far, the default algorithm does not deal with sites in white/black lists? I don't understand what you require to do here. Please elaborate more.

Reply:
That is why I am suggesting to have only a list of workflows that need dedicated data placement (instead of having the site lists a well).
You will, of course, have to modify the standard algorithm such that it can also considers a list of workflow(s) that is retrieved from somewhere else. Other than that, the rest of the logic is already implemented and there is no need to have all this code duplication.

Reply:
@amaltaro , I still don't understand what you are asking for. According to original issue we need to implement the following logic (which is what `_updateSites` function is doing):

2. if a site was added to the SiteBlacklist, that site needs to be removed from the rule ids RSE expression
3. if a site was added to the SiteWhitelist, that site needs to be added to the rule ids RSE expression
4. the new rule ids need to be persisted in the transfer document, replacing those rules that are no longer valid.

And, standard algorithm (which is called in this module as `def execute`) does not do that, it makes a transfer request.  If we not agree how this should be done please let me know your time availability on zoom that we can chat about it.

Reply:
As discussed yesterday, this whole method is probably no longer relevant.